src/Cardano/Wallet/Primitive/AddressDerivation.hs

```diff
@@ -273,6 +285,54 @@ instance
         m <- first (FromMnemonicError . show) (mkMnemonic @mw parts)
         return $ Passphrase $ entropyToBytes $ mnemonicToEntropy m
 
+-- | Encrypt a 'Passphrase' into a format that is suitable for storing on disk
+encryptPassphrase
+    :: MonadRandom m => Passphrase purpose -> m (Hash purpose)
+encryptPassphrase (Passphrase bytes) = do
+    salt <- getRandomBytes @_ @ByteString 16
+    let params = Parameters
+            { iterCounts = 20000
+            , outputLength = 64
+            }
+    return $ Hash $ BA.convert $ mempty
+        <> BS.singleton (fromIntegral (BS.length salt))
+        <> salt
+        <> BA.convert @ByteString (fastPBKDF2_SHA512 params bytes salt)
+
+-- | Check whether a 'Passphrase' matches with a stored 'Hash'
+checkPassphrase
+    :: Passphrase purpose
+    -> Hash purpose
+    -> Either ErrWrongPassphrase ()
+checkPassphrase received stored = do
+    salt <- getSalt stored
+    unless (constantTimeEq (encryptPassphrase received salt) stored) $
+        Left ErrWrongPassphrase
+  where
+    getSalt :: Hash purpose -> Either ErrWrongPassphrase (Passphrase "salt")
+    getSalt (Hash bytes) = do
+        len <- case BS.unpack (BS.take 1 bytes) of
+            [len] -> Right $ fromIntegral len
+            _ -> Left ErrWrongPassphrase
+        Right $ Passphrase $ BA.convert $ BS.take len (BS.drop 1 bytes)
+    constantTimeEq :: Hash purpose -> Hash purpose -> Bool
+    constantTimeEq (Hash a) (Hash b) =
+        BA.convert @_ @ScrubbedBytes a == BA.convert @_ @ScrubbedBytes b
+
+-- | Indicate a failure when checking for a given 'Passphrase' match
+data ErrWrongPassphrase = ErrWrongPassphrase
+    deriving stock (Show, Eq)
+
+-- | Little trick to be able to provide our own random "salt" in order to
```

Comment:
I found this very confusing at first.

> our own random "salt"

why call it random then?

```suggestion
-- | Little trick to be able to provide our own "random" salt in order to
```

or maybe:

```suggestion
-- | Little trick to provide a specific salt instead of a random one in order to
```

**Regardless**, adding the type signature as FYI in this comment somehow would make the point more explicit / difficult to miss:

```haskell
>>> encryptPassphrase
	    :: MonadRandom m => Passphrase purpose -> m (Hash purpose)
>>> encryptPassphrase pwd (Passphrase @"salt" salt)
Hash "..."
```
if you don't find it too ugly.

Also, interesting trick 😮

Reply:
I am all for better comments :+1: